docs/aspnetwebapi-facility.md

```diff
@@ -0,0 +1,139 @@
+
+### Self Hosted
+
+This is a WebApi project that is served up using self hosting capabilities or with OWIN self hosting. Typically something
+that runs in a windows service outside of IIS.
+
+## What do I need to set it up?
+
+You need to determine whether your webapi is hosted in or outside of IIS. Then pull down the `Castle.AspNet.WebApiFacility` 
+into your project from NuGet. 
+
+### IIS or WebHosted
+
+```csharp
+public class MvcApplication : System.Web.HttpApplication
+{
+	protected void Application_Start()
+	{
+		AreaRegistration.RegisterAllAreas();
+		GlobalConfiguration.Configure(WebApiConfig.Register);
+		FilterConfig.RegisterGlobalFilters(GlobalFilters.Filters);
+		RouteConfig.RegisterRoutes(RouteTable.Routes);
+		BundleConfig.RegisterBundles(BundleTable.Bundles);
+
+		var container = new WindsorContainer();
+		container.Register(Component.For<WebApiService>().LifestyleScoped()); // <- `Per Web Request`
+		container.Register(Component.For<WebApiController>().LifestyleScoped()); // <- `Per Web Request`
+
+		container.AddFacility<AspNetWebApiFacility>(x => x.WithLifestyleScopedPerWebRequest());
+	}
+}
+```
+
+The scoped lifestyles are an optional extra if you are using scopes in your web project, they will emulate a `Per Web Request` 
+lifestyle. If not, you can simply register your controllers as transient and they will be implicitly lifestyled for the 
+duration of a web request but behave as transients normally do for services consumed by controllers.
+
+### Self Hosted
+
+Here you will not need anything radically different from the webhosting scenario. You will however need to apply
+the `UsingSelfHosting` method to your facility on startup. 
+
+Here is a rather comprehensive example of how you could achieve this using OWIN. The same is also possible with vanilla
+webapi self hosting. 
+
+```csharp
+public interface IAnyService
+{
+	string Get();
+}
+
+public class AnyService : IAnyService
+{
+	public string Get()
+	{
+		return Guid.NewGuid().ToString("N");
+	}
+}
+
+public class ValuesController : ApiController
+{
+	private readonly IAnyService _anyService;
+
+	public ValuesController(IAnyService anyService)
+	{
+		_anyService = anyService;
+	}
+
+	// GET api/values 
+	public IEnumerable<string> Get()
+	{
+		return new string[] { "value1", "value2", _anyService.Get() };
+	}
+}
+
+public class Startup
+{
+	public void Configuration(IAppBuilder appBuilder)
+	{
+		HttpConfiguration config = new HttpConfiguration();
+
+		config.MapHttpAttributeRoutes();
+
+		config.Routes.MapHttpRoute(
+			name: "DefaultApi",
+			routeTemplate: "api/{controller}/{id}",
+			defaults: new { id = RouteParameter.Optional }
+		);
+
+		appBuilder.UseWebApi(config);
+
+		var container = new WindsorContainer();
+
+		container.Register(Component.For<ValuesController>().LifestyleScoped()); // <- `Per Web Request`
+		container.Register(Component.For<IAnyService>().ImplementedBy<AnyService>().LifestyleScoped()); // <- `Per Web Request`
```

Comment:
With the change to using scopes for per request how does that work if you also use scopes inside your requests. e.g. if `IAnyService` is marked scoped and it get resolved while you are in a nested scope inside the "per request" scope won't it get released on disposal of the nested scope not the "per request" scope?

Having a very quick look at the Autofac documentation they appear to "tag" scopes so resolution inside that nested scope will know to step outwards to find the scope with the same tag that it was registered with, their [`InstancePerRequest` is really an `InstancePerMatchingLifetimeScope`](https://github.com/autofac/Autofac/blob/4a75106b7f7dbf0d62f5452f3c3f0d58333df414/src/Autofac/RegistrationExtensions.cs#L1400-L1408) with a tag.

Reply:
New child scope = new instance for LifestyleScoped. Things resolved after the nested scope has been disposed for the same web request also give you the same instance you had before outside of the child scope. I caution against the usage of this because if you don't also release then there is no guarantee that it will get disposed. The other reason I think this is rather yukky is because you would have to inject IWindsorContainer to get your hands on the BeginScope component registration method, this is possibly making your class vulnerable to implementing Service Location which just sounds like a bad idea. Microsoft took this one step further and implemented GetService on IDependencyScope while implementing IDispose for the .NET Framework. They made this a first class citizen on the dependency injection abstractions and I don't think it ended very well.

Reply:
I would have thought some people would want to use some sort of unit of work scope inside the web request?

Reply:
I think you need to model your problem and use types for this pattern so you fallback on to the compiler. Tagging stuff moves a compilation problem to the run-time.  Not a fan.

Reply:
UnitOfWork is definitely possible. Can post an example on a separate repo if you like?

Reply:
A UoW example would be good, still unclear how it works because Windsor treats both scopes as just scopes and wouldn't know one service is meant to be per request even if you have multiple UoW scopes run during the lifetime of the request.

I wasn't suggesting users directly tag their scopes. With Autofac it is an implementation detail so service registration can be tied to a specific scope with `InstancePerRequest()` so resolution steps up looking for a scope with the same tag it'll resolve inside that scope rather than the most nested scope.

Reply:
Yes, the idea of walking up and out to a known scope. Let me work through some of the changes and get back to you on this one. I will create a new repo demonstrating this, and then we can discuss it and decide on refinements.

Reply:
@jonorossi - I have added a very rough example of something [here](https://github.com/fir3pho3nixx/Windsor.NestedScopes).

Once you clone it dont forget to run:

```
git submodule update --init --recursive
```

Please checkout the [HomeController](https://github.com/fir3pho3nixx/Windsor.NestedScopes/blob/master/Web/Controllers/HomeController.cs#L29).

Reply:
Example output: 

![image](https://user-images.githubusercontent.com/607402/32749130-47ac042c-c8b6-11e7-8793-6a5a021c1bd2.png)

Reply:
@jonorossi - ^^ Ping ^^

Reply:
I'll take a look at this tomorrow.